pallets/nft-marketplace/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+		#[pallet::weight(10_000_000)]
+		pub fn add(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+			currency: CurrencyOf<T>,
+			amount: BalanceOf<T>,
+		) -> DispatchResult {
+			let seller = ensure_signed(origin.clone())?;
+
+			// Check that the seller is the owner of the nft
+			ensure!(
+				Self::is_owner(seller.clone(), class_id, instance_id)?,
+				Error::<T>::NotOwner,
+			);
+
+			// Check that the nft is not yet for sale
+			ensure!(
+				!<Gallery<T>>::contains_key(class_id, instance_id),
+				Error::<T>::AlreadyForSale
+			);
+
+			// Transfer the NFT to the parachain account
+			T::NonFungibles::transfer(&class_id.into(), &instance_id.into(), &Self::account())
+				.map_err(|_| Error::<T>::FailedNftTransfer)?;
+
+			// Put the nft for sale
+			let sale = Sale {
+				seller,
+				price: AskingPrice { currency, amount },
+			};
+			<Gallery<T>>::insert(class_id, instance_id, sale.clone());
+			Self::deposit_event(Event::ForSale(sale));
+
+			Ok(())
+		}
+
+		/// Remove an NFT from sale
+		///
+		/// The sellers of an NFT that is for sale can call this extrinsic to reclaim
+		/// ownership over their NFT and remove it from sale.
+		///
+		/// Fails if
+		///   - the nft is not for sale
+		///   - `origin` is not the seller of the NFT
+		///   - transferring the ownership of the NFT back to the seller fails
+		#[pallet::weight(10_000_000)]
+		pub fn remove(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			let sale = <Gallery<T>>::get(class_id, instance_id).ok_or(Error::<T>::NotForSale)?;
+
+			// Ensure that the buyer is not the seller of the NFT
+			ensure!(who == sale.seller, Error::<T>::NotOwner);
+
+			// Transfer the NFT back to the seller, i.e., the original owner of this NFT
+			T::NonFungibles::transfer(&class_id.into(), &instance_id.into(), &sale.seller)
+				.map_err(|_| Error::<T>::FailedNftTransfer)?;
+
+			// Remove the NFT from the gallery
+			<Gallery<T>>::remove(class_id, instance_id);
+			Self::deposit_event(Event::Removed {
+				class_id,
+				instance_id,
+			});
+
+			Ok(())
+		}
+
+		/// Buy the given nft
+		///
+		/// Fails if
+		///   - the NFT is not for sale
+		///   - `origin` is the seller of the NFT
+		///   - `origin` does not have enough balance of the currency the nft is being sold in
+		///   - transferring the asking price from the buyer to the seller fails
+		///   - transferring the nft to the buyer fails
+		#[pallet::weight(10_000_000)]
+		#[transactional]
+		pub fn buy(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+		) -> DispatchResult {
+			let buyer = ensure_signed(origin.clone())?;
+			let sale = <Gallery<T>>::get(class_id, instance_id).ok_or(Error::<T>::NotForSale)?;
+
+			// Ensure that the buyer is not the seller of the NFT
+			ensure!(buyer != sale.seller, Error::<T>::IsSeller);
```

Comment:
What's the reason for this? It's easy enough to generate a new account for self-dealing, so it seems like a bit of a meaningless check that just costs an extra branch in the common case

Reply:
The idea here really was to avoid having a seller buy their own NFT, as that wouldn't make much sense and I'd see that as something we could prevent. But you have a point and this is also something that the UI could be user-friendly about and not necessarily something the "protocol" needs to block 👍